kythe/java/com/google/devtools/kythe/platform/shared/CompilationUnits.java

```diff
@@ -0,0 +1,162 @@
+  private static final byte[] TAG_ENV = "ENV".getBytes(CHARSET);
+
+  /**
+   * Computes the standard digest for the specified {@link CompilationUnit} in stringified
+   * hexidecimal form.
+   */
+  public static String digestFor(CompilationUnit compilationUnit) {
+    CompilationUnit unit = canonicalize(compilationUnit);
+    ByteArrayDataOutput w = ByteStreams.newDataOutput();
+    putv(w, TAG_CU, compilationUnit.getVName());
+    for (FileInput ri : compilationUnit.getRequiredInputList()) {
+      putv(w, TAG_RI, ri.getVName());
+      put(w, TAG_IN, ri.getInfo().getPath(), ri.getInfo().getDigest());
+    }
+    put(w, TAG_ARG, unit.getArgumentList());
+    put(w, TAG_OUT, unit.getOutputKey());
+    put(w, TAG_SRC, unit.getSourceFileList());
+    put(w, TAG_CWD, unit.getWorkingDirectory());
+    put(w, TAG_CTX, unit.getEntryContext());
+    for (Env env : unit.getEnvironmentList()) {
+      put(w, TAG_ENV, env.getName(), env.getValue());
+    }
+    for (Any d : unit.getDetailsList()) {
+      w.write(TAG_DET);
+      w.writeByte(10);
+      w.write(d.getTypeUrl().getBytes(CHARSET));
+      w.writeByte(0);
+      w.write(d.getValue().toByteArray());
+      w.writeByte(0);
+    }
+    return DIGEST.hashBytes(w.toByteArray()).toString();
+  }
+
+  private static void putv(ByteArrayDataOutput w, byte[] tag, VName v) {
+    put(w, tag, v.getSignature(), v.getCorpus(), v.getRoot(), v.getPath(), v.getLanguage());
+  }
+
+  private static void put(ByteArrayDataOutput w, byte[] tag, String... vals) {
+    put(w, tag, Arrays.asList(vals));
+  }
+
+  private static void put(ByteArrayDataOutput w, byte[] tag, Iterable<String> vals) {
+    w.write(tag);
+    w.writeByte(10);
+    for (String val : vals) {
+      w.write(val.getBytes(CHARSET));
+      w.writeByte(0);
+    }
+  }
+
+  private static int compareInputs(FileInput left, FileInput right) {
+    int r = left.getInfo().getDigest().compareTo(right.getInfo().getDigest());
+    if (r != 0) {
+      return r;
+    }
+    return left.getInfo().getPath().compareTo(right.getInfo().getPath());
+  }
+
+  private static List<FileInput> sortAndDedupe(List<FileInput> orig) {
+    if (orig.size() <= 1) {
+      return orig;
+    }
+    List<FileInput> ri = Lists.newArrayList(orig);
+    Collections.sort(ri, (a, b) -> a.getInfo().getDigest().compareTo(b.getInfo().getDigest()));
+    // Invariant: All elements at or before i are unique (no duplicates).
+    int i = 0, j = 1;
+    while (j < ri.size()) {
+      // If ri[j] ≠ ri[i], it is a new element, not a duplicate.  Move it
+      // next in sequence after i and advance i; this ensures we keep the
+      // first occurrence among duplicates.
+      //
+      // Otherwise, ri[k] == ri[i] for i <= k <= j, i.e., we are scanning a
+      // run of duplicates, and should leave i where it is.
+      if (compareInputs(ri.get(i), ri.get(j)) != 0) {
+        i++;
+        if (i != j) {
+          FileInput tmp = ri.get(i);
```

Comment:
I think you just need `ri.set(i, ri.get(j));` without a swap.  If `i != j`, then `i` was already determined to be a dup and can just be overwritten.